### PR TITLE
Feature/build yaml themes

### PR DESCRIPTION
**Pull request by SergiyIteo**
Implement themes generation got from build.yaml

### CODE REVIEW

File: packages/theme_tailor/lib/src/generator/theme_tailor_generator.dart [hunk 1]
```diff
@@ -124,9 +140,24 @@ class ThemeTailorGenerator extends GeneratorForAnnotation<Tailor> {
...
+    List<String> annotationThemes;
+    if (!annotation.read('themes').isNull) {
+      annotationThemes = List<String>.from(
+        annotation.read('themes').listValue.map((e) => e.toStringValue()),
+      );
+    } else {
+      annotationThemes = [];
+    }
+
+    var pubspecThemes = (builderOptions.config['themes'] as YamlList).map((element) => element.toString()).toList();
```
Comment (Axot017) [inline packages/theme_tailor/lib/src/generator/theme_tailor_generator.dart:152]: config is already parsed and it has type `Map<String, dynamic>` so it's simple json not `YamlList `

File: packages/theme_tailor/pubspec.yaml [hunk 2]
```diff
@@ -22,6 +22,7 @@ dependencies:
   source_gen: ^1.2.2
   source_helper: ^1.3.2
   theme_tailor_annotation: ^1.0.1
+  yaml: ^3.1.1
```
Comment (Axot017) [inline packages/theme_tailor/pubspec.yaml:25]: we don't need this dependency

File: packages/theme_tailor/lib/src/generator/theme_tailor_generator.dart (hunk 1 above)
Comment (SergiyIteo) [reply packages/theme_tailor/lib/src/generator/theme_tailor_generator.dart:152]: Ive checked twice and it was YamlList

File: packages/theme_tailor/pubspec.yaml (hunk 2 above)
Comment (SergiyIteo) [reply packages/theme_tailor/pubspec.yaml:25]: It's needed to map YamlList

File: packages/theme_tailor/lib/src/generator/theme_tailor_generator.dart [hunk 3]
```diff
@@ -221,20 +250,17 @@ class _TailorClassASTVisitor extends SimpleAstVisitor {
...
 
     if (fieldType != null && fieldNamesToCheck.contains(fieldName)) {
-      final childTypeEntities =
-          fieldType.childEntities.map((e) => e.toString()).toList();
+      final childTypeEntities = fieldType.childEntities.map((e) => e.toString()).toList();
       if (childTypeEntities.length >= 2 && childTypeEntities[0] == 'List') {
         final typeWithBraces = childTypeEntities[1];
-        fieldTypes[fieldName] =
-            typeWithBraces.substring(1, typeWithBraces.length - 1);
+        fieldTypes[fieldName] = typeWithBraces.substring(1, typeWithBraces.length - 1);
```
[the remark this answers is not in the document]
Comment (SergiyIteo) [reply packages/theme_tailor/lib/src/generator/theme_tailor_generator.dart]: Done

File: packages/theme_tailor/example/lib/build_yaml_themes.dart [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import 'package:flutter/material.dart';
+import 'package:theme_tailor_annotation/theme_tailor_annotation.dart';
+
+part 'build_yaml_themes.tailor.dart';
+
+@tailor
+class $_AnotherTheme {
+  static List<Color> themes = [Colors.amber, Colors.blueGrey.shade800];
```
Comment (Axot017) [inline packages/theme_tailor/example/lib/build_yaml_themes.dart:8]: @Rongix Shouldn't we throw error if list is shorter than number of themes?

Comment (Rongix) [reply packages/theme_tailor/example/lib/build_yaml_themes.dart:8]: We should throw an error. I will create a task for it

File: packages/theme_tailor/pubspec.yaml (hunk 2 above)
Comment (SergiyIteo) [reply packages/theme_tailor/pubspec.yaml]: Removed the dependency

File: packages/theme_tailor/example/build.yaml [hunk 5]
```diff
@@ -0,0 +1,6 @@
+targets:
+  $default:
+    builders:
+      theme_tailor:
+        options:
+          themes: ["light", "dark", "superDark"]
```
Comment (Rongix) [inline packages/theme_tailor/example/build.yaml:6]: It needs a new line

File: packages/theme_tailor/README.md [hunk 6]
```diff
@@ -153,6 +153,8 @@ The generator will create a static getter with a list of the generated themes:
 final allThemes = MyTailorGeneratedTheme.themes;
 ```
 
+If `themes` property is already used in the tailor class, the generator will use another name and print a warning.
+
```
Comment (Rongix) [inline packages/theme_tailor/README.md:157]: It's already on the develop, why it shows as a change here? Can you fix the merge?

File: packages/theme_tailor/CHANGELOG.md [hunk 7]
```diff
@@ -1,17 +1,20 @@
+# 1.0.4
+- Fix name collision when theme class contains `themes` field
```
Comment (Rongix) [inline packages/theme_tailor/CHANGELOG.md:2]: Currently there are more changes in the changelog, also you need to add a changelog entry in the tailor annotation and here that it is possible to load themes from build.yaml

**Review by Rongix — COMMENTED** [review]
Probably something went wrong during your merge because it shows unnecessary changes from previous PRs. Try to fix it please.

Also remember to add proper changelog entry for theme_tailor package about the ability to load themes from build.yaml
Also need to add additional info how to setup themes from build.yaml in the readme

File: packages/theme_tailor/lib/src/generator/theme_tailor_generator.dart [hunk 8]
```diff
@@ -124,9 +140,26 @@ class ThemeTailorGenerator extends GeneratorForAnnotation<Tailor> {
...
+        .map((element) => element.toString())
+        .toList();
+
+    if (annotationThemes.isNotEmpty) {
+      return annotationThemes;
+    } else if (pubspecThemes.isNotEmpty) {
+      return pubspecThemes;
+    } else {
+      return ['light', 'dark'];
+    }
```
Comment (Rongix) [inline packages/theme_tailor/lib/src/generator/theme_tailor_generator.dart:162]: Maybe smth like that. You can for sure make it more pretty

```suggestion
    if (!annotation.read('themes').isNull) {
      return List<String>.from(
        annotation.read('themes').listValue.map((e) => e.toStringValue()),
      );
    }

    var pubThemes = (builderOptions.config['themes'] as List)
        .map((element) => element.toString())
        .toList();
        
    const defaultThemes = ['light', 'dark'];
    
    return pubThemes.isNotEmpty? pubThemes : defaultThemes;
```

File: packages/theme_tailor/example/build.yaml (hunk 5 above)
Comment (Rongix) [reply packages/theme_tailor/example/build.yaml:6]: Up

**Pull request merged**